samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -47,20 +47,20 @@
    * transformed {@link MessageStream}.
    *
    * @param mapFn the function to transform a message to another message
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM> MessageStream<TM> map(MapFunction<M, TM> mapFn);
+  <OM> MessageStream<OM> map(MapFunction<? super M, ? extends OM> mapFn);
 
   /**
    * Applies the provided 1:n function to transform a message in this {@link MessageStream}
    * to n messages in the transformed {@link MessageStream}
    *
    * @param flatMapFn the function to transform a message to zero or more messages
-   * @param <TM> the type of messages in the transformed {@link MessageStream}
+   * @param <OM> the type of messages in the output {@link MessageStream}
    * @return the transformed {@link MessageStream}
    */
-  <TM> MessageStream<TM> flatMap(FlatMapFunction<M, TM> flatMapFn);
+  <OM> MessageStream<OM> flatMap(FlatMapFunction<? super M, ? extends OM> flatMapFn);
```

Comment:
Refer to out discussion earlier on PR #51 , the whole purpose of this PR is to relax the function type parameter, not the MessageStream type parameter. The purpose is to relax the restriction for user to provide their own functions to be applied to the MessageStream&lt;M&gt;. Hence, here are the basic rules I applied in this change:
1. No type change to the types we define (i.e. MessageStream, StreamGraph, OperatorSpec, Window, etc)
2. Relaxed the type parameters of functions user supplies s.t. the user supplied functions can be applied to the message types in MessageStream.

That's why {code}? super M{code} and {code}? extends M{code} are defined as the type parameters for the functions, not to the MessageStream class. Restricting the type to M totally defeat the whole purpose of this PR.